manager/controlapi/ca_rotation.go

```diff
@@ -0,0 +1,216 @@
+				urls = append(urls, c.URL)
+			}
+		}
+		externalCA.UpdateURLs(urls...)
+		crossSignedCert, err = externalCA.CrossSignRootCA(ctx, newRootCA)
+	}
+
+	if err != nil {
+		log.G(ctx).WithError(err).Error("unable to generate a cross-signed certificate for root rotation")
+		return grpc.Errorf(codes.Internal, "unable to generate a cross-signed certificate for root rotation")
+	}
+
+	cluster.RootCA.RootRotation = &api.RootRotation{
+		CACert:            rootCert,
+		CAKey:             rootKey,
+		CrossSignedCACert: crossSignedCert,
+	}
+	cluster.RootCA.LastForcedRotation = version
+	return nil
+}
+
+// If there is no desired CA key, ensure that there is at least one external CA that uses the desired CA cert, if
+// a desired CA cert is specified.  If the desired CA cert is not specified, and there is no currently root CA key,
+// ensure that the config continues to provide an external CA corresponding to the cluster CA cert.
+func validateExternalCA(cluster *api.Cluster) error {
+	config := cluster.Spec.CAConfig
+
+	if len(config.SigningCACert) > 0 && len(config.SigningCAKey) == 0 {
+		for _, ca := range cluster.Spec.CAConfig.ExternalCAs {
+			if ca.Protocol == api.ExternalCA_CAProtocolCFSSL && bytes.Equal(config.SigningCACert, ca.CACert) {
+				return nil
+			}
+		}
+		return grpc.Errorf(codes.InvalidArgument, "no desired CA signing key nor external CA corresponding to the desired CA signing cert")
+	}
+
+	// there is no need to check for the root rotation CA cert, because if there was a root rotation CA cert without a root rotation key,
+	// it would have had to have been specified in the cluster spec's CAConfig.
+
+	if len(config.SigningCACert) == 0 && len(cluster.RootCA.CACert) > 0 && len(cluster.RootCA.CAKey) == 0 {
+		for _, ca := range cluster.Spec.CAConfig.ExternalCAs {
+			// support older cluster specs that did not specify a ca cert for the external CA
+			if ca.Protocol == api.ExternalCA_CAProtocolCFSSL && (len(ca.CACert) == 0 || bytes.Equal(cluster.RootCA.CACert, ca.CACert)) {
+				return nil
+			}
+		}
+		return grpc.Errorf(codes.InvalidArgument, "no current CA signing key nor external CA corresponding to the current CA signing cert")
+	}
+
+	return nil
+}
+
+// validateAndUpdateCA validates a cluster's desired CA configuration spec and maybe updates the cluster's RootCA object
+func validateAndUpdateCA(ctx context.Context, securityConfig *ca.SecurityConfig, oldConfig *api.CAConfig, cluster *api.Cluster) error {
+	if err := validateExternalCA(cluster); err != nil {
+		return err
+	}
+
+	newConfig := &cluster.Spec.CAConfig
+	// The force rotate flag, if it changes, must monotonically increase
+	if oldConfig.ForceRotate > newConfig.ForceRotate {
+		return grpc.Errorf(codes.InvalidArgument, "the ForceRotate value must be nondecreasing")
+	}
+
+	if len(newConfig.SigningCAKey) > 0 && len(newConfig.SigningCACert) == 0 {
```

Comment:
Another case where we could have a helper to not force me to think what this means 😬 

Arguably a `CanSign` or a `LocalSigner` would validate this inconsistency.